core/src/main/java/tc/oc/pgm/map/source/GitMapSourceFactory.java

```diff
@@ -0,0 +1,152 @@
+package tc.oc.pgm.map.source;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+import org.eclipse.jgit.api.CloneCommand;
+import org.eclipse.jgit.api.FetchCommand;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.ResetCommand;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.api.errors.TransportException;
+import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider;
+import tc.oc.pgm.api.map.MapSource;
+import tc.oc.pgm.api.map.exception.MapMissingException;
+
+public class GitMapSourceFactory extends SystemMapSourceFactory {
+
+  private final String gitURI;
+  private final Logger logger;
+  private final File dir;
+  // The factory needs to parse the source differently depending on if credentials are provided
+  private static boolean credentialsProvided;
+  private boolean updateOnRestart;
+
+  private CredentialsProvider provider = null;
+
+  private static File getFile(String source) {
+    credentialsProvided = isCredentialsProvided(source);
+
+    int uniqueIdentifier =
+        Math.abs(source.hashCode()); // Prevents errors on similar repository names
+
+    String dirName;
+    if (credentialsProvided)
+      dirName =
+          (source.substring(source.lastIndexOf("/"), source.indexOf(";")) + "_" + uniqueIdentifier);
+    else dirName = (source.substring(source.lastIndexOf("/") + 1) + "_" + uniqueIdentifier);
+
+    return new File("./gitMaps/" + dirName);
```

Comment:
Make the default folder `maps/url`

If my repository is `github.com/Electroid/maps`, then the folder will be lowercased to: `maps/github.com/electroid/maps`

Reply:
that both creates a mess to navigate thru and adds severe inconsistencies from what characters are allowed in URLs and what is allowed on folder names. 
Trying to remove invalid characters will lead to possible collissions.

I think it's best to give them an id. in Cardinal we used the hash of the url for the folder name. Not very human readable, but it's the plugin handling those folders anyways, and it can't have these issues.

Reply:
A hash is not really user-friendly. If too many folders is a concern, maybe `github-com-electroid-maps` would be better? `.`, `/`, and `_` become `-`.

Reply:
It would be easier to read like that, but you still have collision issues.
github.com/Electroid/2-maps
github.com/Electroid-2/maps

Those would both end up with the same folder, and i don't think there's any good way to resolve it. Would have to start designing some "folder-name-escaping" technique where `-` becomes `--` and so on, wich feels like a bad solution in the long run

Reply:
Right now i name folders the last word in url + a hash like
`Tournament_651324`

I could make something like electroids
`github-com-electroid-maps`
but then i would have to check for duplicate folders and do some escaping technique like Pablo mentioned.
I could also do Electroids format and add the unique hash at the end.

I could also support custom names in the config, so i could just prompt the user that if they are trying to download some repo that will crash name wise they just need to change the name of either one? I think maybe custom folder names would be nice anyways.

I will move repos to `maps` folder.

Reply:
I'd rather do the `-` and accept the risk of collision (which should be extremely low).